### PR TITLE
Added Import From RapidPro Dialog

### LINKED ISSUE #7829: Create a new dialog "Import from Rapidpro"
This is related to the PLH work, and relevant for the GPI Retreat taking place on 12th September.

image

The "Import from Rapidpro" dialog:

* There are two buttons at the top "Get user data" and "Get flow data"
* `Site:` label, followed by an input box which takes a string
* `Token:` label, followed by a box that leads to a new dialog (see below)
* Checkbox, default unchecked, with the label "Flatten data"
* Checkbox, default unchecked, with the label "Set start date". Clicking on the checkbox displays an input box where the user can input the date.
* Checkbox, default unchecked, with the label "Set end date". Clicking on the checkbox displays an input box where the user can input the date.
* If either "Set start date" or "Set end date" are checked, two new inputs are visible. "Date format:" followed by a drop down of potential date formats given. This works like the date format dropdown in the "Make Date" dialog. And, "Timezone", followed by an input box where the user inputs a string (default "UTC"). 

When pressing the "Token" box, a new dialog pops up.
This is a simple dialog with one label ("Token:") and an input box which takes a string:
When OK is pressed on this sub-dialog, the code `set_rapidpro_key(token = "value")` is run (with `"value"` replaced with the string in the token input box).

Upon pressing OK for "get user data", the following R code is run to import the data

```
get_user_data(rapidpro_site = "VALUE IN SITE BOX", token = get_rapidpro_key(), flatten = checkbox_value, date_from = "input value from set start date checkbox" (default NULL), date_to = "input value from set start date checkbox" (default NULL), format_date = "format_date_value", tzone_date = "timezone_value")
```

For "get flow data", the same R code is run to import the data, but the function is now 
```
get_flow_names
```

### PR DESCRIPTION
**Pull request by anastasia-mbithe**
Fixes #7829 
This is ready @lilyclements , Please test using some real-time data / a valid key.

### CODE REVIEW

**Review by lilyclements — COMMENTED** [review]
Looks great! Two comments:

1. Can you set up the dialogs so that they cannot be Maximised/Minimised, the FixedToolOption and it opens CentreScreen? This can be done in the dialog properties.
2. The token does not read in correctly - you lose the file directory when it is read in and just read in the name of the file

**Comment by lilyclements** [comment]
@anastasia-mbithe I made the changes that I commented about. Can you check you are okay with these? If so, this can then be reviewed.

**Comment by lilyclements** [comment]
Quick problem I just noticed - 

```
# Code generated by the dialog, Import from RapidPro
rapidpror::set_rapidpro_key(key="C:/Users/lclem/OneDrive/Documents/GitHub/ParentText-data-analysis/R Code/tokens/PT_malaysia_key.txt", file=TRUE)

rapidpror::set_rapidpro_site(site="https://app.rapidpro.io/api/v2/")

rapidpror::get_user_data(rapidpro_site=rapidpror::get_rapidpro_site(), token=rapidpror::get_rapidpro_key())

```

We get the data, but do not actually import it into R-Instat. So we need a "New Data Frame Name:" button with an import. @anastasia-mbithe is this something you can sort?

**Comment by lilyclements** [comment]
@anastasia-mbithe I made the changes that I commented about. Can you check you are okay with these? If so, this can then be reviewed. It would be great to have this in for Monday's GPI meeting

**Comment by anastasia-mbithe** [comment]
@lilyclements I think everything is okay now. Just made very few changes.

**Review by lilyclements — COMMENTED** [review]
@anastasia-mbithe looks good! 

@africanmathsinitiative/developers this is ready for review

**Comment by lilyclements** [comment]
https://github.com/orgs/africanmathsinitiative/teams/developers this is ready for review

**Comment by lloyddewit** [comment]
@N-thony please could you (or your delegate) peer review?
thanks

File: instat/dlgImportFromRapidPro.vb [hunk 1]
```diff
@@ -17,5 +17,198 @@
...
+        ucrPnlImportFromRapidPro.AddRadioButton(rdoUserData)
+        ucrPnlImportFromRapidPro.AddRadioButton(rdoFlowData)
+
+        ucrInputRapidProSite.SetParameter(New RParameter("site", 0))
+        ucrInputRapidProSite.SetLinkedDisplayControl(lblRapidProSite)
+
+        ucrChkFlattenData.SetText("Flatten")
+        ucrChkFlattenData.SetParameter(New RParameter("flatten", 2))
+        ucrChkFlattenData.SetValuesCheckedAndUnchecked("TRUE", "FALSE")
+        'ucrChkFlattenData.SetRDefault("FALSE")
```
Comment (N-thony) [inline instat/dlgImportFromRapidPro.vb:55]: can you remove this?

**Review by N-thony — CHANGES_REQUESTED** [review]
@lloyddewit I have also work on it with @anastasia-mbithe perhaps you can peer review? Thanks

File: instat/sdgImportFromRapidPro.vb [hunk 2]
```diff
@@ -15,9 +15,42 @@
...
+        ucrInputTokenPath.SetParameter(clsNewKeyParameter, 0)
     End Sub
-End Class
\ No newline at end of file
+
+    Private Sub cmdChooseFile_Click(sender As Object, e As EventArgs) Handles cmdChooseFile.Click
+        Dim strFileName As String = ""
+        Dim strFileExt As String = ""
+        Dim strTemp As String = ""
+
```
Comment (lloyddewit) [inline instat/sdgImportFromRapidPro.vb:37]:
```suggestion
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@anastasia-mbithe Thank you, this looks good. Just a couple of small suggestions

File: instat/dlgImportFromRapidPro.vb [hunk 3]
```diff
@@ -17,5 +17,197 @@
...
+        dctDateFormat.Add("Month(abbr)/Day/Year(4-digit)", Chr(34) & "%b/%d/%Y" & Chr(34))
+        ucrInputDateFormat.SetItems(dctDateFormat)
+        ucrInputDateFormat.SetDropDownStyleAsEditable(bAdditionsAllowed:=True)
+        ucrInputDateFormat.SetRDefault(Chr(34) & "%Y%m%d" & Chr(34))
+
+        ucrInputTimezone.SetParameter(New RParameter("tzone_date", 6))
+        ucrInputTimezone.SetLinkedDisplayControl(lblTimezone)
+        dctTimezone.Add("UTC", Chr(34) & "UTC" & Chr(34))
+        dctTimezone.Add("EAT", Chr(34) & "EAT" & Chr(34))
+        dctTimezone.Add("GMT", Chr(34) & "GMT" & Chr(34))
```
[the remark this answers is not in the document]
Comment (anastasia-mbithe) [reply instat/dlgImportFromRapidPro.vb:102]: I think it's totally okay to have all time zones included.

[the remark this answers is not in the document]
Comment (anastasia-mbithe) [reply instat/dlgImportFromRapidPro.vb:102]: Generally, time zones are many, so I have just added a few others. Maybe @lilyclements could help us identify the time zones needed.

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@anastasia-mbithe just one more small suggestion, then I'm happy to approve.
Thanks

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern if you can test/approve, then we can merge, thanks

**Review by rdstern — APPROVED** [review]
This all looks fine to me - at least just by clicking.  I don't have a database to check with.  So I guess @lilyclements should also approve.

**Comment by lloyddewit** [comment]
> This all looks fine to me - at least just by clicking. I don't have a database to check with. So I guess @lilyclements should also approve.

@lilyclements approved on 11 Sep and there are no functional changes since then. So I'll go ahead and merge. I hope that's OK.

**Pull request merged**